### ISSUE TITLE
Proposal: `mergeDeep`

### ISSUE THREAD

**Issue by devinrhode2** [MathisBullinger/froebel#15]
Basically I'm not sure about lodash's `_.merge`.

The docs are extremely "wordy" around enumerable properties, giving me pause about the implementation, and the type defs I’ve seen in lodash (so far) seem overly complicated.

Related: https://github.com/sindresorhus/ts-extras/issues/55

**Comment by MathisBullinger** [comment]
Thanks for the suggestion! I think that would make a great addition to the existing utilities. Will try and find some time this weekend to add it to the library.

**Comment by MathisBullinger** [comment]
It's implemented under the name `merge` by bd90ce2dc963fef6b216b087e2efd0098971b9cb and included in `v0.23.0`.

That was way more interesting than expected because of all the special cases that need to be considered.

For example, merging self-referential data:

```js
const a = {}
const b = {}
a.foo = b
b.bar = a
const merged = merge(a, b)
```
What should `merged` be? Clearly `{ foo: merged, bar: merged }`.

But what about this?
```js
const a = [1, 2]
a.push(a)
const b = [3, 4]
b.push(b)
const merged = merge(a, b)
```
Probably `[ 1, 2, merged, 3, 4, merged ]` (that's how it's implemented) though I'd understand if someone argues `merged[2]` and `merged[5]` shouldn't be the same reference.

And what about references in `Map`s?
```js
const a = new Map()
a.set([a, a])
const b = new Map()
b.set([b, b])
const merged = merge(a, b)
```
I don't think there is any right or wrong answer here but I decided self-references in both keys and values will also be self-referencing in the result, so `[...merged.entries()]` will be `[ [merged, merged] ]`.
For some reason in the last example Lodash will return `[ [a, a], [merged, merged] ]` which I don't think makes any sense at all.

There were also some interesting problems in the type signature. E.g. what is the type of `a` when merging `{ a: 1 }` and `{ a?: 2 }`? It could be `1` or `2` or `undefined` (if the second object is `{ a: undefined }`) but it shouldn't be optional because there will always be a value for `a` in the first object, so the resulting type should be `{ a: 1 | 2 | undefined }`.

And what about if one or both of the merged values is a union type? E.g. merging `string | { a: 1; d: 4 }` and `number | { a: 2; c: 3 }`. All the possible types of the unions must be considered so the resulting type has to be `number | { a: 2; c: 3 } | { a: 2; c: 3; d: 4 }`.

And what about something like `merge({ list: [1, 2] }, { list: new Set([2, 3]) })`? Should arrays and `Set`s be merged? Same could be asked for regular objects and `Map`s or other iterable structures. For now I decided to treat them as separate types so if the first parameter is an array and the second one a `Set` then the `Set` will override the array and their values won't be merged. But I think it might make sense to add the option to define custom merge strategies for these cases in the future.
